src/modules/job-manager/plugins/jobspec-default.c

```diff
@@ -0,0 +1,376 @@
+    if (queues) {
+        json_object_foreach (queues, queue, value) {
+            if (get_queue_defaults (h, queue, &queue_defaults, &e) < 0) {
+                errprintf (error,
+                           "Error parsing %s queue policy: %s",
+                           queue,
+                           e.text);
+                return -1;
+            }
+        }
+    }
+    return 0;
+}
+
+/* Build a defaults table that overlays:
+ *  [policy.jobspec.defaults.system]
+ *  [queues.<name>.policy.defaults.system]
+ *
+ * and assign it to 'defaults'.  If no defaults are configured, set defaults
+ * to NULL and return success.
+ * Caller must free the resulting object (N.B. contains refs to conf object).
+ * Returns 0 on success.  On failure, set 'error' and return -1.
+ */
+static int get_policy_defaults (flux_t *h,
+                                json_t *jobspec,
+                                json_t **defaults,
+                                flux_error_t *error)
+{
+    flux_error_t e;
+    json_t *general_defaults;
+    json_t *queue_defaults;
+    json_t *merged;
+    const char *queue;
+
+    if (get_general_defaults (h, &general_defaults, &e) < 0) {
+        errprintf (error, "Error parsing default policy: %s", e.text);
+        return -1;
+    }
+    if (get_queue (jobspec, general_defaults, &queue, &e) < 0) {
+        errprintf (error, "Error parsing default queue name: %s", e.text);
+        return -1;
+    }
+    if (get_queue_defaults (h, queue, &queue_defaults, &e) < 0) {
+        /* If this fails it means 'queue' is not listed in [queues] or
+         * [queues] is missing.  Allow job to proceed for now, if only to
+         * avoid breaking fluxion t1006-qmanager-multiqueue.t and other tests.
+         */
+        queue_defaults = NULL;
+    }
+    if (merge_tables (general_defaults, queue_defaults, &merged) < 0) {
+        errprintf (error, "Out of memory while parsing policy");
+        return -1;
+    }
+    *defaults = merged;
+    return 0;
+}
+
+/* Generate a jobspec-update event context, containing updates
+ * for attributes present in 'defaults' but not 'jobspec'.
+ * If there are no updates, return an empty object.
+ * On error, return NULL with errno set.
+ */
+static json_t *generate_update (json_t *defaults, json_t *jobspec)
+{
+    const char *key;
+    json_t *value;
+    json_t *update;
+    char nkey[128];
+
+    if (!(update = json_object ()))
+        goto nomem;
+    json_object_foreach (defaults, key, value) {
+        json_t *val = jobspec ? json_object_get (jobspec, key) : NULL;
+
+        /* Special: treat user duration=0 as unset per RFC 14.
+         */
+        if (val != NULL
+            && streq (key, "duration")
+            && json_real_value (val) == 0)
+            val = NULL;
+
+        if (val == NULL) {
+            snprintf (nkey, sizeof (nkey), "attributes.system.%s", key);
+
+            /* Special: allow duration to be configured as FSD per RFC 33
```

Comment:
this comment tripped me up, "allow duration" sounds like you're allowing fsd, but what you're really saying is "duration may be configured as FSD", so we have to handle as a special case.

Reply:
Right, I should improve that comment.  It may be _configured_ as an FSD but it has to be a number in jobspec, so we can't just copy it over there like other attributes.